shared/request_caching.go

```diff
@@ -35,7 +36,9 @@ func (w *cachingResponseWriter) Write(data []byte) (int, error) {
 		w.statusCode = http.StatusOK
 	}
 
-	return w.b.Write(data)
+	_, w.err = w.b.Write(data)
+
+	return w.delegate.Write(data)
```

Comment:
What about writing to `delegate` first, and early returning (without writing to buffer) if error occurs?